File: Daily Questions/1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.c
/**
 * @brief Returns the minimum number of swaps to make the string balanced.
 *
 * @param s (char*): The input string consisting of '[' and ']'.
 * @return (int): The minimum number of swaps required to balance the string.
 */
int minSwaps(char* s) {
    int balance = 0;  // Tracks the balance of unmatched opening brackets '['
    int unmatched_close_brackets = 0;  // Tracks unmatched closing brackets ']'

    // Traverse the string until the null terminator
    for (; *s; s++)
        if (balance == 0 && *s == ']')
            // Increment unmatched closing brackets count if we encounter ']' without matching '['
            unmatched_close_brackets++;
        else
            // Increment balance for '[' and decrement for ']'
            balance += 2 * (*s == '[') - 1;

    // Each swap fixes two unmatched brackets, so we need (unmatched_close_brackets + 1) / 2 swaps
    return ++unmatched_close_brackets / 2;
}

File: Daily Questions/1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.py
class Solution:
    def minSwaps(self, s: str) -> int:
        """
        Returns the minimum number of swaps to make the string balanced.

        Parameters:
        s (str): The input string consisting of '[' and ']'.

        Returns:
        int: The minimum number of swaps required to balance the string.
        """
        unmatched_open_brackets = 0

        # Traverse the string and count unmatched open brackets.
        for bracket in s:
            if bracket == '[':
                unmatched_open_brackets += 1
            elif unmatched_open_brackets > 0:
                # Found a matching ']', decrement the unmatched open brackets count.
                unmatched_open_brackets -= 1
        
        # We need (unmatched_open_brackets // 2) swaps to balance all unmatched brackets.
        return (unmatched_open_brackets + 1) // 2
